data/transactions/logic/debugger.go

```diff
@@ -16,294 +16,143 @@
-	Stack        []basics.TealValue `codec:"stack"`
-	Scratch      []basics.TealValue `codec:"scratch"`
-	Error        string             `codec:"error"`
-	OpcodeBudget int                `codec:"budget"`
-	CallStack    []CallFrame        `codec:"callstack"`
-
-	// global/local state changes are updated every step. Stateful TEAL only.
-	transactions.EvalDelta
-}
+	// BeforeTxn is called before a transaction is executed.
+	// groupIndex refers to the index of the transaction in the transaction group that was just executed.
+	BeforeTxn(ep *EvalParams, groupIndex int) error
 
-// GetProgramID returns program or execution ID that is string representation of sha256 checksum.
-// It is used later to link program on the user-facing side of the debugger with TEAL evaluator.
-func GetProgramID(program []byte) string {
-	hash := sha256.Sum256([]byte(program))
-	return hex.EncodeToString(hash[:])
-}
-
-func makeDebugState(cx *EvalContext) *DebugState {
-	disasm, dsInfo, err := disassembleInstrumented(cx.program, nil)
-	if err != nil {
-		// Report disassembly error as program text
-		disasm = err.Error()
-	}
+	// AfterTxn is called after a transaction has been executed.
+	// groupIndex refers to the index of the transaction in the transaction group that was just executed.
+	AfterTxn(ep *EvalParams, groupIndex int) error
 
-	// initialize DebuggerState with immutable fields
-	ds := &DebugState{
-		ExecID:      GetProgramID(cx.program),
-		Disassembly: disasm,
-		PCOffset:    dsInfo.pcOffset,
-		GroupIndex:  int(cx.groupIndex),
-		TxnGroup:    cx.TxnGroup,
-		Proto:       cx.Proto,
-	}
+	// BeforeLogicEval is called before an app or LogicSig is evaluated.
+	BeforeLogicEval(cx *EvalContext) error
 
-	globals := make([]basics.TealValue, len(globalFieldSpecs))
-	for _, fs := range globalFieldSpecs {
-		// Don't try to grab app only fields when evaluating a signature
-		if (cx.runModeFlags&modeSig) != 0 && fs.mode == modeApp {
-			continue
-		}
-		sv, err := cx.globalFieldToValue(fs)
-		if err != nil {
-			sv = stackValue{Bytes: []byte(err.Error())}
-		}
-		globals[fs.field] = stackValueToTealValue(&sv)
-	}
-	ds.Globals = globals
+	// AfterLogicEval is called after an app or LogicSig is evaluated.
+	AfterLogicEval(cx *EvalContext, evalError error) error
 
-	if (cx.runModeFlags & modeApp) != 0 {
-		ds.EvalDelta = cx.txn.EvalDelta
-	}
+	// BeforeTealOp is called before the op is evaluated
+	BeforeTealOp(cx *EvalContext) error
 
-	return ds
-}
+	// AfterTealOp is called after the op has been evaluated
+	AfterTealOp(cx *EvalContext, evalError error) error
 
-// LineToPC converts line to pc
-// Return 0 on unsuccess
-func (d *DebugState) LineToPC(line int) int {
-	if len(d.PCOffset) == 0 || line < 1 {
-		return 0
-	}
+	// BeforeInnerTxnGroup is called before an inner transaction group is executed.
+	// Each inner transaction within the group calls BeforeTxn and subsequent hooks, as described
+	// in the lifecycle diagram.
+	BeforeInnerTxnGroup(ep *EvalParams) error
```

Comment:
Is the inner ep supplied, or the parent?  I'm guessing the inner. Let's document, and rename the argument to either `parent` or `innerEp`.

Reply:
It's the inner. Clarified in 942901c